core/src/main/java/com/datastax/oss/driver/internal/core/auth/PlainTextAuthProviderBase.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.auth;
+
+import com.datastax.oss.driver.api.core.auth.AuthProvider;
+import com.datastax.oss.driver.api.core.auth.Authenticator;
+import com.datastax.oss.driver.api.core.auth.SyncAuthenticator;
+import com.datastax.oss.driver.api.core.metadata.EndPoint;
+import com.datastax.oss.driver.api.core.session.Session;
+import com.datastax.oss.driver.shaded.guava.common.base.Charsets;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Common infrastructure for plain text auth providers.
+ *
+ * <p>This can be reused to write an implementation that retrieves the credentials from another
+ * source than the configuration.
+ */
+@ThreadSafe
+public abstract class PlainTextAuthProviderBase implements AuthProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PlainTextAuthProviderBase.class);
+
+  private final String logPrefix;
+
+  /**
+   * @param logPrefix a string that will get prepended to the logs (this is used for discrimination
+   *     when you have multiple driver instances executing in the same JVM). Built-in
+   *     implementations fill this with {@link Session#getName()}.
+   */
+  protected PlainTextAuthProviderBase(@NonNull String logPrefix) {
+    this.logPrefix = Objects.requireNonNull(logPrefix);
+  }
+
+  /**
+   * Retrieves the credentials from the underlying source.
+   *
+   * <p>This is invoked every time the driver opens a new connection.
+   */
+  @NonNull
+  protected abstract Credentials getCredentials();
+
+  @NonNull
+  @Override
+  public Authenticator newAuthenticator(
+      @NonNull EndPoint endPoint, @NonNull String serverAuthenticator) {
+    return new PlainTextAuthenticator(getCredentials());
+  }
+
+  @Override
+  public void onMissingChallenge(@NonNull EndPoint endPoint) {
+    LOG.warn(
+        "[{}] {} did not send an authentication challenge; "
+            + "This is suspicious because the driver expects authentication",
+        logPrefix,
+        endPoint);
+  }
+
+  @Override
+  public void close() {
+    // nothing to do
+  }
+
+  protected static class Credentials {
+
+    private final String username;
+    private final String password;
```

Comment:
As an additional security measure, I would recommend storing the password as a char array, for the reasons explained [here](https://stackoverflow.com/questions/8881291/why-is-char-preferred-over-string-for-passwords). Converting from `char[]` to `byte[]` later on will be a bit annoying, but it's still doable.

Reply:
Good point, I was not aware of that.

> Converting from char[] to byte[] later on will be a bit annoying

We might as well use `byte[]` directly then, wdyt?

Reply:
I fear that users might be surprised by a password field of type `byte[]`, but I am OK going that direction if you think it's easier.

Reply:
All things considered, I'll stick to `char[]`. Using bytes forces the UTF-8 encoding on the users, that's kind of an implementation detail.